### PR TITLE
[Core][Geometries] Checking distance on local coordinates of Triangle3D3

### PR DESCRIPTION
**Pull request by loumalouomega**
A proper tolerance is considered for the distance in the Triangle3D3

Replaces #4246

### CODE REVIEW

**Comment by pooyan-dadvand** [comment]
But do we want to check if we are on plane?

**Comment by loumalouomega** [comment]
> But do we want to check if we are on plane?

Yes, otherwise we basically we will be doing what is done right now (and then giving false positives)

File: kratos/geometries/triangle_3d_3.h [hunk 1]
```diff
@@ -909,7 +910,53 @@ template<class TPointType> class Triangle3D3
         CoordinatesArrayType point_projected = rPoint;
         const array_1d<double,3> vector_points = rPoint - r_center.Coordinates();
         const double distance = inner_prod(vector_points, normal);
-        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) { // Not in the plane, projecting
+
+        // We check if we are on the plane
+        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) {
```
Comment (RiccardoRossi) [inline kratos/geometries/triangle_3d_3.h]: what are u doing here? i don't think this warning belongs to the master...

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: Which warning?

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h]: Ok, I see

**Comment by pooyan-dadvand** [comment]
> Yes, otherwise we basically we will be doing what is done right now (and then giving false positives)

This is `PointLocalCoordinate` and would only give local coordinates as they are. As I remember we wanted to have the in plane test to be in `IsInside` not this one.

**Comment by loumalouomega** [comment]
> > Yes, otherwise we basically we will be doing what is done right now (and then giving false positives)
> 
> This is `PointLocalCoordinate` and would only give local coordinates as they are. As I remember we wanted to have the in plane test to be in `IsInside` not this one.

I see. Huuum, I think I understand

**Comment by loumalouomega** [comment]
Done @pooyan-dadvand. I needed to update the result of one test in the MeshingApplication. PointLocator was indirectly extrapolating results, this is solved now

File: applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ void NodalValuesInterpolationProcess<TDim>::ExtrapolateValues(
     KDTreeType tree_points(point_list_destination.begin(), point_list_destination.end(), bucket_size);
 
     // We extrapolate the nodes that cannot been found
-    for (auto& p_node : rToExtrapolateNodes) {
+    for (auto p_node : rToExtrapolateNodes) {
```
Comment (loumalouomega) [inline applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp:300]: This is a pointer, so I removed the &

File: kratos/tests/cpp_tests/geometries/test_triangle_3d_3.cpp [hunk 3]
```diff
@@ -172,10 +172,10 @@ namespace Testing
     KRATOS_TEST_CASE_IN_SUITE(Triangle3D3IsInside, KratosCoreGeometriesFastSuite) {
         auto geom = GenerateRightTriangle3D3<NodeType>();
 
-        Point PointInside(0.33, 0.33, 0.0);
+        Point PointInside(0.5 * std::cos(Globals::Pi/4), 0.5, 0.5 * std::sin(Globals::Pi/4));
```
Comment (loumalouomega) [inline kratos/tests/cpp_tests/geometries/test_triangle_3d_3.cpp:175]: These tests were wrong previously, now they are fine

**Comment by loumalouomega** [comment]
I was thinking that it would be consistent to define the same for the Line2D2

File: kratos/geometries/triangle_3d_3.h [hunk 4]
```diff
@@ -867,7 +869,34 @@ template<class TPointType> class Triangle3D3
...
-        PointLocalCoordinates( rResult, rPoint );
+        // The center of the geometry
+        const auto& r_center = this->Center();
+
+        // Tangent vectors
+        const array_1d<double, 3> tangent_xi  = this->GetPoint(1) - this->GetPoint(0);
+        const array_1d<double, 3> tangent_eta = this->GetPoint(2) - this->GetPoint(0);
+
+        // We compute the normal to check the normal distances between the point and the triangles, so we can discard that is on the triangle
+        array_1d<double, 3> normal;
```
Comment (philbucher) [inline kratos/geometries/triangle_3d_3.h]: This is copy-paste of the `Normal` implemented in this class, please remove the duplication by using the function

**Review by philbucher — CHANGES_REQUESTED** [review]
After having dealt with this in detail during the previous days I have to say that this PR makes full sense to me and seems to be correct, thanks @loumalouomega 

I have one minor request but other than that I would approve in a week or so if none has further comments

**Comment by philbucher** [comment]
@ others:
@loumalouomega and I have been working on the Projections, they are easy to use and can be found [here => GeometricalProjectionUtilities](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/geometrical_projection_utilities.h)

**Comment by loumalouomega** [comment]
Done @philbucher

File: kratos/geometries/line_2d_2.h [hunk 5]
```diff
@@ -883,6 +884,31 @@ class Line2D2 : public Geometry<TPointType>
...
+        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) {
+            if (std::abs(distance) > 1.0e-6 * Length()) {
+                KRATOS_WARNING("Line2D2") << "The point of coordinates X: " << rPoint[0] << "\tY: " << rPoint[1] << " it is in a distance: " << std::abs(distance) << std::endl;
+                return false;
+            }
+
+            // Not in the plane, but allowing certain distance, projecting
+            noalias(point_projected) = rPoint - normal * distance;
+        }
+
```
Comment (pooyan-dadvand) [inline kratos/geometries/line_2d_2.h:899]: Is there any test for this? I think the implementation is not correct.

File: kratos/geometries/line_2d_2.h [hunk 6]
```diff
@@ -883,6 +884,31 @@ class Line2D2 : public Geometry<TPointType>
...
+        // We compute the normal to check the normal distances between the point and the triangles, so we can discard that is on the triangle
+        const CoordinatesArrayType aux_point = ZeroVector(3);
+        const array_1d<double, 3> normal = this->UnitNormal(aux_point);
+
+        // We compute the distance, if it is not in the pane we
+        const Point point_to_project(rPoint);
+        Point point_projected;
+        const array_1d<double,3> vector_points = rPoint - r_center.Coordinates();
+
+        const double distance = GeometricalProjectionUtilities::FastProjectDirection(*this, point_to_project, point_projected, normal,  vector_points, 0);
```
Comment (pooyan-dadvand) [inline kratos/geometries/line_2d_2.h]: why you take normal to calculate the projection?

File: kratos/geometries/line_2d_2.h (hunk 5 above)
Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h:899]: Yes, the current tests for IsInside. It is doing the same for the triangle

File: kratos/geometries/line_2d_2.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h]: The normal and the vector pointing to the center. It is the same we have done in the triangle

File: kratos/geometries/line_2d_2.h (hunk 5 above)
Comment (pooyan-dadvand) [reply kratos/geometries/line_2d_2.h:899]: But this is a line_2d_2

Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h:899]: Is the same than the triangle in 3D, it follows the same principles

Comment (pooyan-dadvand) [reply kratos/geometries/line_2d_2.h:899]: No, I disagree, the implmentation can be simplified and seems to be a copy paste of triangle. I suggest to revise it and make test for it.

Comment (philbucher) [reply kratos/geometries/line_2d_2.h:899]: I recommend GeometricalProjectionUtilities::FastProjectOnLine, does exactly that

Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h:899]: Done

Comment (pooyan-dadvand) [reply kratos/geometries/line_2d_2.h:899]: Why the warning? Consider that we will run this methods millions of times!

File: kratos/geometries/line_2d_2.h [hunk 7]
```diff
@@ -883,7 +884,20 @@ class Line2D2 : public Geometry<TPointType>
...
-        PointLocalCoordinates( rResult, rPoint );
+        // We compute the distance, if it is not in the pane we
+        const Point point_to_project(rPoint);
+        Point point_projected;
+        const double distance = GeometricalProjectionUtilities::FastProjectOnLine(*this, point_to_project, point_projected);
+
+        // We check if we are on the plane
+        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) {
+            if (std::abs(distance) > 1.0e-6 * Length()) {
+                KRATOS_WARNING("Line2D2") << "The point of coordinates X: " << rPoint[0] << "\tY: " << rPoint[1] << " it is in a distance: " << std::abs(distance) << std::endl;
```
Comment (philbucher) [inline kratos/geometries/line_2d_2.h]:
```suggestion
                KRATOS_WARNING_FIRST_N("Line2D2", 10) << "The point of coordinates X: " << rPoint[0] << "\tY: " << rPoint[1] << " it is in a distance: " << std::abs(distance) << std::endl;
```

**Comment by loumalouomega** [comment]
@philbucher I extended to the triangle too

**Comment by loumalouomega** [comment]
Ping @philbucher

**Comment by philbucher** [comment]
@pooyan-dadvand do you still have comments?
Otherwise I would approve

**Comment by pooyan-dadvand** [comment]
If there is test for line 2d2 I am ok

**Comment by loumalouomega** [comment]
> If there is test for line 2d2 I am ok

There is one https://github.com/KratosMultiphysics/Kratos/blob/529e69547607627191276743292d931d0356fc75/kratos/tests/cpp_tests/geometries/test_line_2d_2.cpp#L116

**Review by philbucher — APPROVED** [review]
no more comments, approving, thx

**Comment by philbucher** [comment]
I haven't run the tests in the mappingapp which uses this stuff but I think I can adapt if sth fails, anyway working on it now

**Comment by loumalouomega** [comment]
Thanks

File: kratos/geometries/line_2d_2.h [hunk 8]
```diff
@@ -883,7 +884,20 @@ class Line2D2 : public Geometry<TPointType>
...
-        PointLocalCoordinates( rResult, rPoint );
+        // We compute the distance, if it is not in the pane we
+        const Point point_to_project(rPoint);
+        Point point_projected;
+        const double distance = GeometricalProjectionUtilities::FastProjectOnLine(*this, point_to_project, point_projected);
+
+        // We check if we are on the plane
+        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) {
+            if (std::abs(distance) > 1.0e-6 * Length()) {
+                KRATOS_WARNING_FIRST_N("Line2D2", 10) << "The point of coordinates X: " << rPoint[0] << "\tY: " << rPoint[1] << " it is in a distance: " << std::abs(distance) << std::endl;
```
Comment (philbucher) [inline kratos/geometries/line_2d_2.h:895]: found a bug in my code bcs of this :+1:

Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h:895]: :+1: I am doing a PR, because I have to much warnings because of this

**Pull request merged**